File: python/mirascope/ops/_internal/propagation.py
"""Context propagation utilities for distributed tracing."""

from __future__ import annotations

import logging
import os
from collections.abc import Mapping, MutableMapping
from typing import Literal, TypeAlias

from opentelemetry import propagate
from opentelemetry.context import Context
from opentelemetry.propagators.b3 import B3MultiFormat, B3SingleFormat
from opentelemetry.propagators.composite import CompositePropagator
from opentelemetry.propagators.jaeger import JaegerPropagator
from opentelemetry.propagators.textmap import Setter, TextMapPropagator
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

from ..exceptions import ConfigurationError
from .session import SESSION_HEADER_NAME, current_session

logger = logging.getLogger(__name__)

PropagatorFormat: TypeAlias = Literal[
    "tracecontext", "b3", "b3multi", "jaeger", "composite"
]
CarrierValue: TypeAlias = str | list[str]

# Environment variable names
ENV_PROPAGATOR_FORMAT = "MIRASCOPE_PROPAGATOR"
ENV_PROPAGATOR_SET_GLOBAL = "_MIRASCOPE_PROPAGATOR_SET_GLOBAL"

_PROPAGATOR: ContextPropagator | None = None


class _StrSetter(Setter[MutableMapping[str, str]]):
    """Setter that writes string header values into the carrier."""

    def set(self, carrier: MutableMapping[str, str], key: str, value: str) -> None:
        """Set a single header value on the carrier."""
        carrier[key] = value


_STR_SETTER = _StrSetter()


class ContextPropagator:
    """Manages OpenTelemetry context propagation across service boundaries."""

    _propagator: TextMapPropagator

    def __init__(self, *, set_global: bool = True) -> None:
        """Initialize propagator and optionally set as global.

        Propagator format is determined by MIRASCOPE_PROPAGATOR environment variable.
        Defaults to "tracecontext" if not set.

        Args:
            set_global: Whether to set this propagator as the global textmap propagator.

        Raises:
            ConfigurationError: If the propagator format is invalid.
        """
        env_format = os.environ.get(ENV_PROPAGATOR_FORMAT, "tracecontext")
        propagator_name: PropagatorFormat
        if env_format in ("tracecontext", "b3", "b3multi", "jaeger", "composite"):
            propagator_name = env_format
        else:
            error_message = (
                f"Invalid propagator format: {env_format}. "
                f"Valid options: tracecontext, b3, b3multi, jaeger, composite"
            )
            logger.error(error_message)
            raise ConfigurationError(error_message)
        logger.debug(f"Initializing ContextPropagator with format: {propagator_name}")

        match propagator_name:
            case "tracecontext":
                self._propagator = TraceContextTextMapPropagator()
            case "b3":
                self._propagator = B3SingleFormat()
            case "b3multi":
                self._propagator = B3MultiFormat()
            case "jaeger":
                self._propagator = JaegerPropagator()
            case "composite":
                propagators: list[TextMapPropagator] = [
                    TraceContextTextMapPropagator(),
                    B3SingleFormat(),
                    B3MultiFormat(),
                    JaegerPropagator(),
                ]
                self._propagator = CompositePropagator(propagators)

        if set_global:
            should_set_global = os.environ.get(ENV_PROPAGATOR_SET_GLOBAL, "true")
            if should_set_global.lower() == "true":
                propagate.set_global_textmap(self._propagator)
                logger.debug(f"Set {propagator_name} as global textmap propagator")

    def extract_context(self, carrier: Mapping[str, CarrierValue]) -> Context:
        """Extract OTEL context from carrier headers.

        Args:
            carrier: Dictionary containing HTTP headers or similar carrier data.

        Returns:
            Extracted OpenTelemetry context. Returns empty context if extraction fails.
        """
        try:
            context = self._propagator.extract(carrier=carrier)
            logger.debug("Successfully extracted context from carrier")
            return context
        except Exception as exception:
            logger.debug(
                f"Failed to extract context from carrier: {type(exception).__name__}: {exception}"
            )
            return Context()

    def inject_context(
        self,
        carrier: MutableMapping[str, str],
        context: Context | None = None,
    ) -> None:
        """Inject current OTEL context into carrier headers.

        This method also injects session context if one is active, adding the
        SESSION_HEADER_NAME header to the carrier.

        Args:
            carrier: Mutable mapping (e.g., HTTP headers dict) to inject context into.
            context: Optional specific context to inject. If None, uses current context.
        """
        try:
            self._propagator.inject(
                carrier=carrier, context=context, setter=_STR_SETTER
            )
            logger.debug("Successfully injected context into carrier")
        except Exception as exception:
            logger.debug(
                f"Failed to inject context into carrier: {type(exception).__name__}: {exception}"
            )

        session_ctx = current_session()
        if session_ctx is not None:
            carrier[SESSION_HEADER_NAME] = session_ctx.id


def get_propagator() -> ContextPropagator:
    """Get or create the singleton ContextPropagator instance.

    Reads propagator format from MIRASCOPE_PROPAGATOR environment variable.

    Returns:
        The global ContextPropagator instance.
    """
    global _PROPAGATOR
    if _PROPAGATOR is None:
        _PROPAGATOR = ContextPropagator()
    return _PROPAGATOR


def reset_propagator() -> None:
    """Reset the singleton ContextPropagator instance.

    This is primarily useful for testing to ensure a clean state between tests.
    The next call to get_propagator() will create a new instance.
    """
    global _PROPAGATOR
    _PROPAGATOR = None


def extract_context(carrier: Mapping[str, CarrierValue]) -> Context:
    """Extract OTEL context from carrier headers using the global propagator.

    Args:
        carrier: Dictionary containing HTTP headers or similar carrier data.

    Returns:
        Extracted OpenTelemetry context.
    """
    return get_propagator().extract_context(carrier)


def inject_context(
    carrier: MutableMapping[str, str],
    *,
    context: Context | None = None,
) -> None:
    """Inject current OTEL context into carrier headers using the global propagator.

    This function also injects session context if one is active, adding the
    SESSION_HEADER_NAME header to the carrier.

    Args:
        carrier: Mutable mapping (e.g., HTTP headers dict) to inject context into.
        context: Optional specific context to inject. If None, uses current context.
    """
    get_propagator().inject_context(carrier, context)

File: typescript/src/ops/_internal/propagation.ts
/**
 * Context propagation utilities for distributed tracing.
 *
 * Provides helpers for extracting and injecting trace context from/to HTTP headers
 * and other carriers, supporting multiple propagation formats.
 */

import {
  propagation,
  context as otelContext,
  type Context,
  type TextMapGetter,
  type TextMapSetter,
} from "@opentelemetry/api";
import {
  W3CTraceContextPropagator,
  CompositePropagator,
} from "@opentelemetry/core";
import { B3Propagator, B3InjectEncoding } from "@opentelemetry/propagator-b3";
import { JaegerPropagator } from "@opentelemetry/propagator-jaeger";

import type { PropagatorFormat } from "@/ops/_internal/types";

import { SESSION_HEADER_NAME, currentSession } from "@/ops/_internal/session";

let _propagator: ContextPropagator | null = null;

/**
 * Getter for extracting values from a carrier object.
 */
const headerGetter: TextMapGetter<Record<string, string | string[]>> = {
  get(carrier, key) {
    const value = carrier[key];
    if (Array.isArray(value)) {
      return value[0];
    }
    return value;
  },
  /* v8 ignore start - required by TextMapGetter interface but not called by standard propagators */
  keys(carrier) {
    return Object.keys(carrier);
  },
  /* v8 ignore end */
};

/**
 * Setter for injecting values into a carrier object.
 */
const headerSetter: TextMapSetter<Record<string, string>> = {
  set(carrier, key, value) {
    carrier[key] = value;
  },
};

/**
 * Context propagator for distributed tracing.
 *
 * Handles extraction and injection of trace context from/to HTTP headers
 * using various propagation formats (W3C TraceContext, B3, Jaeger).
 *
 * @example
 * ```typescript
 * // Create and use a propagator
 * const propagator = new ContextPropagator();
 *
 * // Extract context from incoming request
 * const ctx = propagator.extractContext(req.headers);
 *
 * // Inject context into outgoing request
 * const headers: Record<string, string> = {};
 * propagator.injectContext(headers);
 * ```
 */
export class ContextPropagator {
  private readonly _format: PropagatorFormat;

  /**
   * Create a new ContextPropagator.
   *
   * @param setGlobal - Whether to set this as the global propagator (default: true).
   * @param format - The propagation format to use. Defaults to MIRASCOPE_PROPAGATOR
   *                 environment variable or 'tracecontext'.
   */
  constructor(
    private readonly setGlobal = true,
    format?: PropagatorFormat,
  ) {
    this._format =
      format ??
      ((process.env.MIRASCOPE_PROPAGATOR as PropagatorFormat) ||
        "tracecontext");
    const propagatorInstance = this.createPropagator(this._format);

    if (this.setGlobal) {
      propagation.setGlobalPropagator(propagatorInstance);
    }
  }

  /**
   * Create a propagator instance for the given format.
   */
  private createPropagator(format: PropagatorFormat) {
    switch (format) {
      case "tracecontext":
        return new W3CTraceContextPropagator();
      case "b3":
        return new B3Propagator({
          injectEncoding: B3InjectEncoding.SINGLE_HEADER,
        });
      case "b3multi":
        return new B3Propagator({
          injectEncoding: B3InjectEncoding.MULTI_HEADER,
        });
      case "jaeger":
        return new JaegerPropagator();
      case "composite":
        return new CompositePropagator({
          propagators: [
            new W3CTraceContextPropagator(),
            new B3Propagator(),
            new JaegerPropagator(),
          ],
        });
      default:
        return new W3CTraceContextPropagator();
    }
  }

  /**
   * Get the propagation format being used.
   */
  get format(): PropagatorFormat {
    return this._format;
  }

  /**
   * Extract context from a carrier (e.g., HTTP headers).
   *
   * @param carrier - The carrier object containing trace context headers.
   * @returns The extracted OpenTelemetry context.
   */
  extractContext(carrier: Record<string, string | string[]>): Context {
    return propagation.extract(otelContext.active(), carrier, headerGetter);
  }

  /**
   * Inject context into a carrier (e.g., HTTP headers).
   *
   * Also injects the Mirascope session ID header if a session is active.
   *
   * @param carrier - The carrier object to inject trace context into.
   * @param context - Optional context to inject. Defaults to active context.
   */
  injectContext(carrier: Record<string, string>, context?: Context): void {
    propagation.inject(context ?? otelContext.active(), carrier, headerSetter);

    // Also inject session ID if available
    const session = currentSession();
    if (session) {
      carrier[SESSION_HEADER_NAME] = session.id;
    }
  }
}

/**
 * Get the global ContextPropagator instance.
 *
 * Creates one if it doesn't exist.
 *
 * @returns The global ContextPropagator.
 */
export function getPropagator(): ContextPropagator {
  if (!_propagator) {
    _propagator = new ContextPropagator();
  }
  return _propagator;
}

/**
 * Reset the global propagator.
 *
 * Primarily useful for testing.
 */
export function resetPropagator(): void {
  _propagator = null;
}

/**
 * Extract context from a carrier using the global propagator.
 *
 * @param carrier - The carrier object containing trace context headers.
 * @returns The extracted OpenTelemetry context.
 *
 * @example
 * ```typescript
 * // In an HTTP handler
 * const ctx = extractContext(req.headers);
 * ```
 */
export function extractContext(
  carrier: Record<string, string | string[]>,
): Context {
  return getPropagator().extractContext(carrier);
}

/**
 * Inject context into a carrier using the global propagator.
 *
 * @param carrier - The carrier object to inject trace context into.
 * @param context - Optional context to inject. Defaults to active context.
 *
 * @example
 * ```typescript
 * // Before making an outgoing HTTP request
 * const headers: Record<string, string> = {};
 * injectContext(headers);
 * fetch(url, { headers });
 * ```
 */
export function injectContext(
  carrier: Record<string, string>,
  context?: Context,
): void {
  getPropagator().injectContext(carrier, context);
}

/**
 * Run a function with context extracted from a carrier.
 *
 * Extracts trace context from the carrier and runs the function with that
 * context active, ensuring proper parent-child span relationships.
 *
 * @param carrier - The carrier object containing trace context headers.
 * @param fn - The function to execute within the extracted context.
 * @returns The result of the function.
 *
 * @example
 * ```typescript
 * // In an HTTP handler
 * await propagatedContext(req.headers, async () => {
 *   // Spans created here are linked to the incoming trace
 *   await span('handle-request', async (s) => {
 *     // Process request
 *   });
 * });
 * ```
 */
export async function propagatedContext<T>(
  carrier: Record<string, string | string[]>,
  fn: () => T | Promise<T>,
): Promise<T> {
  const ctx = extractContext(carrier);
  return otelContext.with(ctx, () => fn());
}
